mlflow/server/auth/__init__.py

```diff
@@ -8,15 +8,52 @@
 """
 
 import logging
+from flask import Flask, request, make_response, Response, redirect, flash, render_template_string
 
 from mlflow.server import app
-
+from mlflow.server.auth import store
+from mlflow.server.auth.config import app_config
+from mlflow.server.handlers import _get_rest_path
 
 _logger = logging.getLogger(__name__)
 
 
+class ROUTES:
+    USERS = _get_rest_path("/mlflow/users")
+    SIGNUP = _get_rest_path("/mlflow/signup")
+
+
+UNPROTECTED_ROUTES = [ROUTES.USERS, ROUTES.SIGNUP]
+
+
+def is_unprotected_route(path: str) -> bool:
+    if path.startswith("/static"):
+        return True
+    return path in UNPROTECTED_ROUTES
+
+
+def make_basic_auth_response() -> Response:
+    res = make_response()
+    res.status_code = 401
+    res.headers["WWW-Authenticate"] = 'Basic realm="mlflow"'
+    return res
+
+
 def _before_request():
-    # TODO: Implement authentication
+    _user = request.authorization.username if request.authorization else None
+    _logger.info(f"before_request: {request.method} {request.path} (user: {_user})")
+    if is_unprotected_route(request.path):
+        return
+
+    if request.authorization is None:
+        return make_basic_auth_response()
+
+    username = request.authorization.username
+    password = request.authorization.password
```

Comment:
If auth is enabled, we should force `flask` to enable SSL mode, and client must check SSL mode first otherwise reject sending requests with password, otherwise the raw password is transmitted over network without any protection, this is very dangerous.

Reply:
makes sense. Is it sufficient to just check the environment variables `MLFLOW_TRACKING_INSECURE_TLS` and `MLFLOW_TRACKING_SERVER_CERT_PATH`?

Reply:
@WeichenXu123 Do we need to enable SSL on the flask app? For example, you could enable SSL on your server, not the app hosted in that server.

Reply:
> you could enable SSL on your server, not the app hosted in that server.

Yes we should enable SSL on server side, but I meant we should add checking in client side, if client try to connect to the server that is not configured with SSL (i.e.. we can check scheme not starts with "https"), then client should raise error instead of sending requests containing credential to server

Reply:
@WeichenXu123 Raising an error is a breaking change. I think we can emit a warning message. What do you think?